syntax = "proto3";

package resdb.contract;

import "service/contract/proto/func_params.proto";
import "service/contract/proto/contract.proto";
import "service/contract/proto/account.proto";

message Response {
int32 ret = 1;
optional Account account = 2;
optional Contract contract = 3;
optional string res = 4;
}

message RWS {
  enum Type {
    READ = 0;
    WRITE = 1;
  };
  
  Type type = 1;
  bytes address = 2;
  bytes value = 3;
  int32 version = 4;
}

message Request {
    enum CMD {
        NONE = 0;
        CREATE_ACCOUNT = 1; // deploy contract
        DEPLOY = 2; // deploy contract
        EXECUTE = 3; // execute contract
    };

    CMD cmd = 1;
    string caller_address = 2;
    optional DeployInfo deploy_info = 3;
    optional string contract_address = 4;
    optional Params func_params = 5;
    optional Response resp = 7;
    int32 id = 8;
    repeated RWS rws = 9;
}

message ResultInfo{
    repeated RWS rws = 1;
    optional Response resp = 2;
    int32 idx = 3;
}

